Edit project display name & description via settings page

benjaminapetersen opened the pull request:
Adds a drop down menu to top left of settings page (similar to what we have elsewhere) with edit & delete options.  Edit will display an inline form allowing user to update the project's `displayName` and `description`.

![screen shot 2016-01-14 at 4 42 52 pm](https://cloud.githubusercontent.com/assets/280512/12338777/5c39ff2a-badf-11e5-9c28-38c6b200e500.png)
![screen shot 2016-01-15 at 3 01 08 pm](https://cloud.githubusercontent.com/assets/280512/12363813/d827ee9e-bb98-11e5-8d7e-5fd5cb842df5.png)


@jwforres @spadgett @sg00dwin for review!

spadgett commented:
The margin above the first editable field feels a little tight.

Should we use a small textarea for description to be consistent with the New Project form?

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12354314/15c27368-bb63-11e5-8567-83272928d90b.png)

spadgett commented:
https://trello.com/c/wr0L2wW5

spadgett commented:
The margins after the edit and delete menu icons are different in your screenshot.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
-      </div>
+    <div class="breadcrumb"></div>
+    <alerts alerts="alerts"></alerts>
+    <div class="tile__">
```
Underscores accidental?

benjaminapetersen replied:
No, disabled/enable the 'tile'.  I was considering using a tile like we don on DCs and other things, but im not sure it makes sense for settings.  I will remove, unless anyone thinks we should keep it.

Screenshot attached for comparison

![screen shot 2016-01-15 at 9 54 52 am](https://cloud.githubusercontent.com/assets/280512/12355977/21e657f8-bb6e-11e5-9c45-dd28ccd631a8.png)

benjaminapetersen replied:
Agree about the margin above the edit field, though it keeps the 'Display name:' label from jumping when editing.

spadgett commented on assets/app/scripts/controllers/settings.js:
```diff
@@ -26,7 +26,31 @@ angular.module('openshiftConsole')
-        $scope.project = project;
+        angular.extend($scope, {
+          project: project,
+          show: { display: true, form: false, actions: true },
```
Looks like `show.actions` is no longer used.

I suggest one scope var for tracking whether we're showing the form rather than 3 we have to keep in sync.

benjaminapetersen replied:
Yeah, I will eliminate.  It made sense when i had the edit icons, as i could return the regular view but keep the buttons hidden until the promise resolution.  Now that the 'edit' action is in the menu, no longer needed.

spadgett commented on assets/app/scripts/controllers/settings.js:
```diff
@@ -26,7 +26,31 @@ angular.module('openshiftConsole')
+            $scope.toggle();
+            $scope.show.actions = false;
+            ProjectsService
+              .update($routeParams.project, angular.copy(project))
```
We should copy the project before the user begins editing. Otherwise, if the user never clicks save, any cached project object will have wrong data.

Not sure offhand if we ever cache projects, but we might someday.

spadgett replied:
We'll also show the updated values even when the update fails as written, correct?

benjaminapetersen replied:
Nice, I'll move the copy up out of just the `.update()` & make these changes

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+            <li>
+              <a href=""
+                role="button"
+                ng-click="toggle()"
```
So if I select Edit again while editing the form, it toggles back to read only? Probably better to disable the menu.

jwforres commented:
Yes, make description a textarea

On Friday, January 15, 2016, Sam Padgett <notifications@github.com> wrote:

> The margin above the first editable field feels a little tight.
>
> Should we use a small textarea for description to be consistent with the
> New Project form?
>
> [image: openshift_web_console]
> <https://cloud.githubusercontent.com/assets/1167259/12354314/15c27368-bb63-11e5-8567-83272928d90b.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6667#issuecomment-171963337>.
>

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+              <a href=""
+                role="button"
+                ng-click="toggle()"
+                ><i class="fa fa-pencil fa-fw" aria-hidden="true"></i>Edit</a>
```
I think the difference in margins is because there is no space before text.

Does it look better as `<i class="fa fa-pencil fa-fw" aria-hidden="true"></i> Edit`?

benjaminapetersen replied:
Agree, was just looking at DC page to compare, is just the inline space, not CSS.  fixed.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+              <div row>
+                <div flex></div>
+                <div flex grow="2">
+                  <button class="btn btn-default" ng-click="update()">Save</button>
```
Probably should disable the button until one of the fields has changed.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+                      <i class="icon icon-pencil"></i>
+                    </a>
+                  </div>
+                  -->
```
Let's remove the commented-out code if we're going with the menu.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+                         type="text"
+                         name="project_display_name"
+                         placeholder="project display name"
+                         ng-model="project.metadata.annotations['openshift.io/display-name']">
```
We need to remove the annotation entirely if display name is empty.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+                         type="text"
+                         name="project_description"
+                         placeholder="project description"
+                         ng-model="project.metadata.annotations['openshift.io/description']">
```
Same comment here: we should remove the annotation if no value.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+              <div
+                row
+                class='form-group'
+                ng-class="{'has-error': form.project_display_name.$error.oscKeyValid}">
```
When is `$error.oscKeyValid` set? Copy/paste mistake?

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,236 @@
+              <div
+                row
+                class='form-group'
+                ng-class="{'has-error': form.project_description.$error.oscKeyValid}">
```
Same question on `oscKeyValid` as above.

spadgett commented:
Can't decide if we should have a Cancel button. @jwforres ?

benjaminapetersen commented:
@spadgett im inclined toward a cancel button, makes sense!

jwforres commented:
always provide an escape hatch :)

On Friday, January 15, 2016, Sam Padgett <notifications@github.com> wrote:

> Can't decide if we should have a Cancel button. @jwforres
> <https://github.com/jwforres> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6667#issuecomment-171974444>.
>

benjaminapetersen commented:
@spadgett @jwforres fixes pushed.  I used an intermediate model called `$scope.editableFields` to deal with rollbacks. It seems fairly clean.  The only grumble I currently have is the hack in place to make the edit button disabled while editing the form.  Our styles are for a link (which cannot be disabled) so I'm swapping with a disabled button.... :unamused:

spadgett commented on assets/app/scripts/services/projects.js:
```diff
@@ -1,19 +1,30 @@
+                              return memo;
+                            }, {});
+          return resource;
+      };
```
Hm, I'm a little worried about potentially removing other empty annotations we don't know about. I'd suggest only removing display name and description to be safe.

benjaminapetersen replied:
Ok I'm good with that.  Will update.

jwforres replied:
Editing any other annotations on a namespace would cause an authorization
failure
On Jan 15, 2016 3:50 PM, "Sam Padgett" <notifications@github.com> wrote:

> In assets/app/scripts/services/projects.js
> <https://github.com/openshift/origin/pull/6667#discussion_r49903434>:
>
> > +
> > +      // eliminates empty Annotations
> > +      // could potentially move this into a shared Util service
> > +      var cleanAnnotations = function(resource) {
> > +          resource
> > +            .metadata
> > +            .annotations = _.reduce(
> > +                            resource.metadata.annotations,
> > +                            function(memo, val, key) {
> > +                              if(val) {
> > +                                memo[key] = val;
> > +                              }
> > +                              return memo;
> > +                            }, {});
> > +          return resource;
> > +      };
>
> Hm, I'm a little worried about potentially removing other empty
> annotations we don't know about. I'd suggest only removing display name and
> description to be safe.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6667/files#r49903434>.
>

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,119 +1,233 @@
+                ng-if="show.editing"
+                type="button"
+                class="btn btn-link"
+                DISABLED>
```
Does this change how the menu item looks?

The `delete-link` directive has a disabled option. We should add it to `edit-link` and implement them consistently (either styling links or using buttons).

benjaminapetersen replied:
I'll dig into that.  This isn't a directive because it doesn't really encapsulate anything special like the delete button does.  Let me see if there is some nifty Angular thing I missed for this.

benjaminapetersen replied:
Yeah, I can borrow the `ng-class="{ 'disabled-link': show.editing }"` from delete-link, which looks disabled but you can still click it...

benjaminapetersen replied:
Just tested. The delete button also can still be clicked, though it appears to be disabled.  Are we ok with that?

spadgett replied:
You can click the link, but the delete dialog shouldn't appear. Is that what you're seeing? The menu does disappear, which is a bit odd.

I'm OK switching to buttons for disabled links if they look the same with `btn-link` styling. Maybe we should be using buttons in the first place if they can be used inside menus, probably better for accessibility. It might be worth pulling this into a directive for reuse.

benjaminapetersen replied:
Right, I'm preferring that disabled wouldn't even let you click, rather than a click that does nothing.

Buttons can be used inside menus, but unfortunately we lose the hover effect:

![screen shot 2016-01-18 at 1 25 03 pm](https://cloud.githubusercontent.com/assets/280512/12399298/1f707172-bde7-11e5-828b-12de504102d3.png)

"Edit" is a link, will give the blue outline.  "Edit 2" is a button... will not outline.  

Silly that this isn't nicely handled by bootstrap menu. :unamused:  I'll tinker a bit.

benjaminapetersen replied:
Changed, simplified it down to this: 

```javascript
            <li>
              <a
                href=""
                role="button"
                ng-click="toggleEdit(true)"
                ng-class="{ 'disabled-link': show.editing }">
                <i class="fa fa-pencil fa-fw" aria-hidden="true"></i> Edit
              </a>
            </li>
```
`toggleEdit(true)` will just not do anything if it is already editing, so its basically the same markup as the delete directive.  I'd make a directive, but I don't think we gain anything here since there isn't anything reusable (unlike the delete button which is encapsulating the delete modal).

spadgett commented on assets/app/scripts/controllers/settings.js:
```diff
@@ -26,7 +26,71 @@ angular.module('openshiftConsole')
+          canSubmit: function(bool) {
+            $scope.actions.canSubmit = bool;
+          },
+          toggleEdit: function(bool) {
```
`setEditing`? No longer a toggle really.

benjaminapetersen replied:
done

spadgett commented on assets/app/scripts/controllers/settings.js:
```diff
@@ -26,7 +26,71 @@ angular.module('openshiftConsole')
+              return true;
+            }
+            $scope.canSubmit(false);
+            return false;
```
Might be able to handle this is the view with

```
ng-enabled="form.project_display_name.$dirty || form.project_description.$dirty"
```

benjaminapetersen replied:
Good call!  I was able to simplify it a bit further with this:

```html
<button
   class="btn btn-default"
   ng-click="update()"
   ng-disabled="editForm.$pristine">Save</button>
```

Eliminated `.validateSubmit()`

spadgett replied:
even better, thanks!

spadgett commented:
Does the enter key submit the form?

benjaminapetersen commented:
Yup, it does. Was just looking at all of our forms. It appears that Angular supports this out of the box (with inputs, not with textarea).

spadgett commented:
[test]

spadgett commented:
@benjaminapetersen can you look at the test failures? they might be legitimate

benjaminapetersen commented:
will do

benjaminapetersen commented:
should `hack/test-assets.sh` run integration tests as well as our unit tests?

spadgett commented:
> should hack/test-assets.sh run integration tests as well as our unit tests?

I think it was intentional to leave that out... @jwforres

You can run the integration tests with

```
cd assets
grunt test-integration --suite rest_api
```

jwforres commented:
Yes intentional

hack/test-assets.sh should never require the api server to be running
and this one should never require a node be running, only a master:

grunt test-integration --suite rest_api

which means don't have expectations about pods getting scheduled.

On Tuesday, January 19, 2016, Sam Padgett <notifications@github.com> wrote:

> should hack/test-assets.sh run integration tests as well as our unit tests?
>
> I think it was intentional to leave that out... @jwforres
> <https://github.com/jwforres>
>
> You can run the integration tests with
>
> cd assets
> grunt test-integration --suite rest_api
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6667#issuecomment-172903155>.
>

benjaminapetersen commented:
Cool.  Would we ever want something like:

```
hack/test-assets-unit.sh
hack/test-assets-e2e.sh
hack/test-assets-integration.sh
hack/test-assets-all.sh 
```
Thinking about symmetry/workflow when creating a PR.  Maybe we don't need all of that, but I may suggest we rename `test-assets` now that it only runs certain tests.

jwforres commented:
Renaming is fine with me, you'd have to update the references to it in
vagrant-openshift and in the assets/README.md

I wouldn't change it as part of this PR though, open a tech-debt issue and
we can fix it later in the sprint.

On Tuesday, January 19, 2016, Ben Petersen <notifications@github.com> wrote:

> Cool. Would we ever want something like:
>
> hack/test-assets-unit.sh
> hack/test-assets-e2e.sh
> hack/test-assets-integration.sh
> hack/test-assets-all.sh
>
> Thinking about symmetry/workflow when creating a PR. Maybe we don't need
> all of that, but I may suggest we rename test-assets now that it only
> runs certain tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6667#issuecomment-172916175>.
>

benjaminapetersen commented:
sounds good!

benjaminapetersen commented:
@spadgett @jwforres fixed the tests related to the project settings page.  I need to squash, but left the commit separate as I updated our JSHint task to run against `/test/integration` as well & fixed a bunch of items.  

I updated the `it('should delete a project')` test to use the concept of a "Page Object".  There is a new `/integration/page_objects/settings.js` file.  Essentially this abstracts the selector noise out of our tests & makes the flow more terse/readable (in theory).  I may open an issue to update all of our tests to use Page Objects.

[TEST]

spadgett commented:
With your refactoring, how do the errors look when a test fails? Do I see the selector it's using to find an element in the error message? My worry is if we abstract too much in our test code it will be hard to debug failures.

I'm also not entirely convinced this

```
 +          settingsPage
 +            .dropdown()
 +            .click()
 +            .then(function() {
 +              settingsPage
 +                .deleteProjectButton()
 +                .click()
 +                .then(function() {
 +                    settingsPage
 +                      .deleteModal()
 +                      .confirm()
 +                      .click()
 +                      .then(function() {
 +                        h.waitFor(h.presenceOf(settingsPage.alerts.deletionSuccess(), 'Delete failed'));
 +                      });
 +                });
 +            });
```

is more readable than this

```
 -          element(by.css(".action-button .fa-trash-o")).click();
 -          element(by.cssContainingText(".modal-dialog .btn", "Delete")).click();
 -          h.waitForPresence(".alert-success", "marked for deletion");
```

benjaminapetersen commented:
I'd like to reduce that even more, still does seem too noisy in the actual `project.js` test file.  
I can force fail some tests and see how the errors look & get back to you on that.

jwforres commented:
agree with @spadgett, I am not convinced yet

benjaminapetersen commented:
@jwforres @spadgett moving that to an issue for lengthier discussion.  I don't think its quite right, but seems to be worth exploring a bit more.

I'm going to cherry pick that out of here & fix the test w/basic selectors for now.

benjaminapetersen commented:
Updated

benjaminapetersen commented:
[TEST]

spadgett commented:
Did we decide about prefixing test-only CSS classes?

benjaminapetersen commented:
We did not come to a conclusion in that discussion.  I'm assuming it will be something we handle/update in future PRs?

spadgett commented:
> We did not come to a conclusion in that discussion. I'm assuming it will be something we handle/update in future PRs?

My only concern is that it's going to be hard to identify the test-only classes later on. Much easier to do it as you add them.

benjaminapetersen commented:
@spadgett @jwforres since this relates to our "test only" classes discussion....

I think the right way to deal with this is a better abstraction:

```html
<a class="project-edit">Edit</a>
```
```scss
.project-edit {
fa fa-pencil fa-fw
}
```

sg00dwin commented:
@benjaminapetersen I don't mind abstracting these into mixins. Just make sure to include the rules for ``fa-fw`` to ensure a list of icons and text are aligned properly.

```
a.edit:before {
    content: "\f040";
    display: inline-block;
    font-family: fontawesome;
    width: 1.285714em;
    text-align: center;
}
```

benjaminapetersen commented:
@sg00dwin yeah, I'm not going to do that in this PR, but think its worth discussing.  I really mentioned in in relation to [issue #6746](https://github.com/openshift/origin/issues/6746) where we can break tests if we change icons.  It would probably be good to create a separate issue for the less mixin refactor.

benjaminapetersen commented:
[TEST]

spadgett commented:
@benjaminapetersen It looks like you have quota details in the settings page twice after rebasing.

benjaminapetersen commented:
@spadgett updated, all the icons next to headings are now removed, correct?

![screen shot 2016-01-27 at 9 52 09 am](https://cloud.githubusercontent.com/assets/280512/12617007/c770ac84-c4db-11e5-8bac-761504eca3da.png)

benjaminapetersen commented:
[TEST]

spadgett commented:
@benjaminapetersen Yeah, that looks right. You might create a quota to make sure things look OK.

https://docs.openshift.org/latest/dev_guide/quota.html#sample-resource-quota-file

Run `oc create -f resource-quota.json - n my-new-project` as cluster admin.

spadgett commented on assets/app/views/settings.html:
```diff
@@ -1,34 +1,53 @@
+          <a href=""
+            class="dropdown-toggle resource-actions-dropdown"
+            data-toggle="dropdown">
+            <i class="fa fa-ellipsis-v"></i>
```
`aria-hidden="true"`

benjaminapetersen replied:
got it

spadgett commented on assets/app/views/settings.html:
```diff
@@ -175,6 +199,6 @@ <h3 class="text-center">Memory</h3>
         </div>
       </div>
-    </div>
+
```
Hm, did we lose a div?

benjaminapetersen replied:
Yup... `build-assets.sh` wouldn't build w/o that.  I broke the 'edit settings' chunk out into a partial include to make it a bit easier to work through all the tags.  Tempted to do the same with the other components as well...

spadgett replied:
Ah, I see: line 3 is the start tag that got removed.

The `h1` above and everything after seems to have extra indentation.

benjaminapetersen replied:
fixed.

spadgett commented on assets/app/views/_settings-general-info.html:
```diff
@@ -0,0 +1,92 @@
+    row
+    class='form-group'
+    style="margin-top: 10px">
+    <div flex><strong>Display name: </strong></div>
```
Should have an ID and `aria-labelledby` on the input below for accessibility. Or make it a `label` with a `for` attribute.

Here and below for description.

benjaminapetersen replied:
nice, done.

spadgett commented:
Looks good, just a few small accessibility comments.

benjaminapetersen commented:
[TEST]

spadgett commented on assets/test/integration/helpers.js:
```diff
@@ -78,3 +80,15 @@ exports.goToPage = function(uri) {
+
+exports.presenceOf = function(obj, msg) {
+  msg = msg || '';
+  return protractor.ExpectedConditions.presenceOf(obj);
```
Are we not using `msg`?

benjaminapetersen replied:
I think it is supposed to be:
```javascript

exports.presenceOf = function(obj) {
  return protractor.ExpectedConditions.presenceOf(obj);
};

// example:
//  h.waitFor(h.presenceOf(page.header()))
exports.waitFor = function(item, timeout, msg) {
  timeout = timeout || 5000;
  msg = msg || '';
  return browser.wait(item, timeout, msg);
};
```
But for the most part I put the tests back to the old way.  I'll update this quick even though its not being used (until future tests).

benjaminapetersen commented:
First run was a success, second a failure.  Weird.

spadgett commented:
LGTM, [merge]
